### ISSUE TITLE
ByteBuffer contract broken in public API

### ISSUE THREAD

**Issue by revintec** [luben/zstd-jni#78]
https://github.com/luben/zstd-jni/blob/83820d8f241e80848917cc93eda09c28523a040f/src/main/java/com/github/luben/zstd/Zstd.java#L759-L761

this is the latest source as of writting this issue
clearly, dstBuff is passed-in from outside, so its position should be non-zero after the call and thus requires a byteBuffer.flip() by the caller after this function returns, but it's not

the following code demonstrates two problems:
1) line12~13: inconsistent public API, considering the source code above, I think it's a BUG
2) line19: ZstdDictTrainer fails if there is only one sample

![image](https://user-images.githubusercontent.com/8952734/46915516-58d7e280-cfdf-11e8-9126-e9946883b812.png)
output is
why? java.lang.RuntimeException: Error (generic)
java.nio.DirectByteBuffer[pos=28 lim=256 cap=256]
java.nio.DirectByteBuffer[pos=0 lim=25 cap=256]

```Java
class BugReport{
    public static void compress(ByteBuffer input,ZstdDictCompress dict){
        ByteBuffer output=allocateDirect(0x100);
        if(dict!=null)Zstd.compress(output,input,dict);
        else Zstd.compress(output,input,18);
        System.out.println(output);
    }
    public static void main(String... args)throws Exception{
        byte[]data="123456abcdefg123456".getBytes();
        ZstdDictTrainer trainer=new ZstdDictTrainer(0x1000,0x1000);
        trainer.addSample(data);try{trainer.trainSamples();}catch(Throwable e){System.out.println("why? "+e);}

        trainer.addSample(data);byte[]bytes=trainer.trainSamples();
        ByteBuffer src=allocateDirect(0x100);src.put(data).flip();

        compress(src.duplicate(),null);
        compress(src.duplicate(),new ZstdDictCompress(bytes,18));
    }
}
```

**Comment by luben** [comment]
Yes, I see. There are a few places with the same mistake with regards the ByteBuffer contract.

I will take a look on the dict trainer problem first.

**Comment by luben** [comment]
Looks like one sample is not enough for the dict trainer to make a dictionary. It can still make a dictionary in "legacy" mode. I think this has changed in the last Zstd version where a new dict builder was introduced.

**Comment by revintec** [comment]
ok, thx

**Comment by luben** [comment]
I am holding off the release as 1.3.7 is around the corner. See also https://github.com/facebook/zstd/issues/1357

**Comment by revintec** [comment]
@luben thx, I'll try it soon
